File: prototype/rails/src/app/domain/orders/order_payment_test.rb
require "minitest/autorun"
require_relative "order_payment"

class OrderPaymentTest < Minitest::Test
  Payment = Domain::Orders::OrderPayment
  Status = Domain::Orders::OrderStatus

  def test_an_order_awaiting_payment_takes_a_card
    assert Payment.awaits_card?(Status::AWAITING_PAYMENT)
  end

  def test_a_declined_order_takes_another_card
    assert Payment.awaits_card?(Status::PAYMENT_FAILED)
  end

  def test_an_unverified_order_takes_no_card_yet
    refute Payment.awaits_card?(Status::PENDING_VERIFICATION)
  end

  def test_a_paid_order_takes_no_card
    refute Payment.awaits_card?(Status::PAID)
  end

  def test_an_unverified_order_is_unpaid
    assert Payment.unpaid?(Status::PENDING_VERIFICATION)
  end

  def test_a_declined_order_is_unpaid
    assert Payment.unpaid?(Status::PAYMENT_FAILED)
  end

  def test_a_shipped_order_is_not_unpaid
    refute Payment.unpaid?(Status::SHIPPED)
  end

  def test_a_verified_purchaser_pays_an_order_awaiting_payment
    assert Payment.payable?(Status::AWAITING_PAYMENT, true)
  end

  def test_an_unverified_purchaser_pays_nothing
    refute Payment.payable?(Status::AWAITING_PAYMENT, false)
  end

  def test_a_verified_purchaser_cannot_pay_a_delivered_order
    refute Payment.payable?(Status::DELIVERED, true)
  end
end

File: prototype/php/src/app/Domain/Orders/OrderPaymentTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Orders;

it('awaits payment only before a successful charge', function (OrderStatus $status, bool $expected): void {
    expect(OrderPayment::awaitsPayment($status))->toBe($expected);
})->with([
    'pending verification awaits payment' => [OrderStatus::PendingVerification, true],
    'awaiting payment awaits payment' => [OrderStatus::AwaitingPayment, true],
    'payment failed still awaits payment' => [OrderStatus::PaymentFailed, true],
    'paid no longer awaits payment' => [OrderStatus::Paid, false],
    'shipped no longer awaits payment' => [OrderStatus::Shipped, false],
    'delivered no longer awaits payment' => [OrderStatus::Delivered, false],
    'cancelled no longer awaits payment' => [OrderStatus::Cancelled, false],
]);

it('is payable only by a verified purchaser', function (): void {
    expect(OrderPayment::isPayableBy(OrderStatus::PendingVerification, true))->toBeTrue()
        ->and(OrderPayment::isPayableBy(OrderStatus::PendingVerification, false))->toBeFalse();
});

it('is not payable once paid, even by a verified purchaser', function (): void {
    expect(OrderPayment::isPayableBy(OrderStatus::Paid, true))->toBeFalse();
});
